File: ocaml/tennis1.ml

let show_score p1Score p2Score =
    let score = ref "" in
    let tempScore = ref 0 in
    if p1Score = p2Score then
    begin
        match p1Score with
        | 0 -> score := "Love-All"
        | 1 -> score := "Fifteen-All"
        | 2 -> score := "Thirty-All"
        | _ -> score := "Deuce"
    end
    else if p1Score >= 4 || p2Score >= 4 then
    begin
        let minusResult = p1Score - p2Score in
        if minusResult = 1 then score := "Advantage player1"
        else if minusResult = -1 then score := "Advantage player2"
        else if minusResult >= 2 then score := "Win for player1"
        else score := "Win for player2"
    end
    else
    begin
        for i = 1 to 2 do
            if i = 1 then tempScore := p1Score
            else (score := !score ^ "-"; tempScore := p2Score);
            match !tempScore with
            | 0 -> score := !score ^ "Love"
            | 1 -> score := !score ^ "Fifteen"
            | 2 -> score := !score ^ "Thirty"
            | 3 -> score := !score ^ "Forty"
            | _ -> assert false
        done
    end;
    !score

File: python/tennis1.py
class TennisGame1:
    def __init__(self, player1_name, player2_name):
        self.player1_name = player1_name
        self.player2_name = player2_name
        self.p1points = 0
        self.p2points = 0

    def won_point(self, player_name):
        if player_name == "player1":
            self.p1points += 1
        else:
            self.p2points += 1

    def score(self):
        result = ""
        temp_score = 0
        if self.p1points == self.p2points:
            result = {
                0: "Love-All",
                1: "Fifteen-All",
                2: "Thirty-All",
            }.get(self.p1points, "Deuce")
        elif self.p1points >= 4 or self.p2points >= 4:
            minus_result = self.p1points - self.p2points
            if minus_result == 1:
                result = "Advantage player1"
            elif minus_result == -1:
                result = "Advantage player2"
            elif minus_result >= 2:
                result = "Win for player1"
            else:
                result = "Win for player2"
        else:
            for i in range(1, 3):
                if i == 1:
                    temp_score = self.p1points
                else:
                    result += "-"
                    temp_score = self.p2points
                result += {
                    0: "Love",
                    1: "Fifteen",
                    2: "Thirty",
                    3: "Forty",
                }[temp_score]
        return result
